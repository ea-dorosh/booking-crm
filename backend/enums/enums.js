
const customerNewStatusEnum = {
  existing: 0,
  new: 1,
};

const salutationEnum = {
  male: 0,
  female: 1,
};

module.exports = {
  customerNewStatusEnum,
  salutationEnum,
};