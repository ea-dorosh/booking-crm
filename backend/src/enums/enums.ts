export enum CustomerNewStatusEnum {
  Existing = 0,
  New = 1,
}

export enum SalutationEnum {
  Male = 0,
  Female = 1,
}

export enum AppointmentStatusEnum {
  Active = 0,
  Canceled = 1,
}

export enum CustomerSourceEnum {
  Manual = 0,
  Booking = 1,
}

export enum PaymentMethodEnum {
  PayInPerson = 0,
  ApplePay = 1,
  CreditDebitCards = 2,
  Giropay = 3,
  OnlineBankTransfer = 4,
}

export enum PaymentStatusEnum {
  Declined = 0,
  Refunded = 1,
  Successful = 2,
}

export enum InvoiceStatusEnum {
  Paid = 1,
}

// Default sorting constants
export const DEFAULT_APPOINTMENT_SORT_FIELD = `date` as const;
export const DEFAULT_SORT_DIRECTION = `desc` as const;

export enum CategoryStatusEnum {
  Active = `active`,
  Archived = `archived`,
  Disabled = `disabled`,
}

export enum EmployeeStatusEnum {
  Active = `active`,
  Archived = `archived`,
  Disabled = `disabled`,
}

// Feature flags (temporary hardcoded switches)
export const FEATURE_FLAGS = {
  employeeSchedulePeriods: true,
};
