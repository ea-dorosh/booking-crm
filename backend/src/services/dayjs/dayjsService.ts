import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat.js';
import customParseFormat from 'dayjs/plugin/customParseFormat.js';
import isSameOrBefore from 'dayjs/plugin/isSameOrBefore.js';
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter.js';
import duration from 'dayjs/plugin/duration.js';
import isBetween from 'dayjs/plugin/isBetween.js';
import timezone from 'dayjs/plugin/timezone.js';
import utc from 'dayjs/plugin/utc.js';
import weekday from 'dayjs/plugin/weekday.js';
import updateLocale from 'dayjs/plugin/updateLocale.js';
import 'dayjs/locale/de';

dayjs.extend(advancedFormat);
dayjs.extend(customParseFormat);
dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(duration);
dayjs.extend(isSameOrBefore);
dayjs.extend(isSameOrAfter);
dayjs.extend(isBetween);
dayjs.extend(weekday);
dayjs.extend(updateLocale);

dayjs.locale(`de`);
dayjs.updateLocale(`de`, {
  weekStart: 1
});

export { dayjs };