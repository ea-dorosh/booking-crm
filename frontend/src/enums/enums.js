
const customerNewStatusEnum = {
  existing: 0,
  new: 1,
};

const salutationEnum = {
  male: 0,
  female: 1,
};

const appointmentStatusEnum = {
  active: 0,
  canceled: 1,
};

const categoryStatusEnum = {
  active: 'active',
  archived: 'archived',
  disabled: 'disabled',
};

export {
  appointmentStatusEnum,
  customerNewStatusEnum,
  salutationEnum,
  categoryStatusEnum,
};