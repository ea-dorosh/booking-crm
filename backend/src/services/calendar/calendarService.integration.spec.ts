/**
 * Calendar Service Integration Tests
 *
 * Комплексные интеграционные тесты для защиты от регрессий при добавлении новых фич.
 * Эти тесты гарантируют, что существующая логика календаря не сломается.
 *
 * Проверяемые блокировки:
 * - Google Calendar Events
 * - Saved Appointments
 * - Pause Times
 * - Blocked Times
 *
 * Проверяемые интервалы: 15, 30, 60 минут
 * Проверяемые сценарии: 1 и 2 сервиса, несколько работников
 */

import { describe, it, expect } from '@jest/globals';
import {
  calculateEmployeeDayAvailability,
  type EmployeeWorkingDayPure,
  type NormalizedAppointmentPure,
} from './calendarUtils.pure.js';
import { TimeslotIntervalEnum } from '@/enums/enums.js';
import { dayjs } from '@/services/dayjs/dayjsService.js';
import type { Date_ISO_Type, Time_HH_MM_SS_Type } from '@/@types/utilTypes.js';

describe(`Calendar Service Integration Tests`, () => {
  const TIMEZONE = `Europe/Berlin`;
  // Используем будущую дату чтобы избежать проблем с currentTime
  const FUTURE_DATE = `2025-12-15`;

  describe(`✅ Core Functionality: calculateEmployeeDayAvailability`, () => {
    it(`should calculate available slots for employee with 15min interval`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 18:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Fifteen),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 07:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      expect(result.employeeId).toBe(15);
      expect(Number(result.timeslotInterval)).toBe(15);
      expect(result.availableTimes).toBeInstanceOf(Array);
      expect(result.availableTimes.length).toBeGreaterThan(0);
    });

    it(`should calculate available slots for employee with 60min interval`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 1,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Sixty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      expect(result.employeeId).toBe(1);
      expect(result.availableTimes.length).toBeGreaterThan(0);
    });
  });

  describe(`✅ Blocker Sources: All types should be processed`, () => {
    it(`should process Saved Appointments`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 20:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Thirty),
      };

      const blockers: NormalizedAppointmentPure[] = [
        {
          employeeId: 15,
          dateISO: FUTURE_DATE as Date_ISO_Type,
          startTimeMs: dayjs.tz(`${FUTURE_DATE} 10:00`, TIMEZONE).valueOf(),
          endTimeMs: dayjs.tz(`${FUTURE_DATE} 11:00`, TIMEZONE).valueOf(),
        },
      ];

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 07:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        blockers,
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // Should successfully process without errors
      expect(result).toBeDefined();
      expect(result.employeeId).toBe(15);
    });

    it(`should process Pause Times`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [
          {
            startPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 12:00`, TIMEZONE).valueOf(),
            endPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 13:00`, TIMEZONE).valueOf(),
          },
        ],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Thirty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // Should successfully process pause times
      expect(result).toBeDefined();
      expect(result.employeeId).toBe(15);
    });

    it(`should process multiple blocker types together`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 20:00`, TIMEZONE).valueOf(),
        pauseTimes: [
          {
            startPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 12:00`, TIMEZONE).valueOf(),
            endPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 13:00`, TIMEZONE).valueOf(),
          },
        ],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Fifteen),
      };

      const blockers: NormalizedAppointmentPure[] = [
        // Saved Appointment
        {
          employeeId: 15,
          dateISO: FUTURE_DATE as Date_ISO_Type,
          startTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
          endTimeMs: dayjs.tz(`${FUTURE_DATE} 10:00`, TIMEZONE).valueOf(),
        },
        // Google Event
        {
          employeeId: 15,
          dateISO: FUTURE_DATE as Date_ISO_Type,
          startTimeMs: dayjs.tz(`${FUTURE_DATE} 14:00`, TIMEZONE).valueOf(),
          endTimeMs: dayjs.tz(`${FUTURE_DATE} 15:00`, TIMEZONE).valueOf(),
        },
        // Blocked Time
        {
          employeeId: 15,
          dateISO: FUTURE_DATE as Date_ISO_Type,
          startTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
          endTimeMs: dayjs.tz(`${FUTURE_DATE} 18:00`, TIMEZONE).valueOf(),
        },
      ];

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 07:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        blockers,
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // Should handle all blocker types without errors
      expect(result).toBeDefined();
      expect(result.employeeId).toBe(15);
      expect(result.availableTimes).toBeInstanceOf(Array);
    });
  });

  describe(`✅ Timeslot Intervals: Different intervals work correctly`, () => {
    it(`should handle 15 minute intervals`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Fifteen),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      expect(Number(result.timeslotInterval)).toBe(15);
      expect(result.availableTimes.length).toBeGreaterThan(0);
    });

    it(`should handle 30 minute intervals`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 14,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Thirty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      expect(Number(result.timeslotInterval)).toBe(30);
      expect(result.availableTimes.length).toBeGreaterThan(0);
    });

    it(`should handle 60 minute intervals`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 1,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Sixty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      expect(Number(result.timeslotInterval)).toBe(60);
      expect(result.availableTimes.length).toBeGreaterThan(0);
    });
  });

  describe(`✅ Edge Cases: Boundary conditions`, () => {
    it(`should respect advance booking time`, () => {
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `02:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Thirty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 10:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // First slot should be at least 2 hours from current time
      if (result.availableTimes.length > 0) {
        const firstSlot = dayjs(result.availableTimes[0].minPossibleStartTimeMs).tz(TIMEZONE);
        const minBookingTime = dayjs(currentTimeMs).add(2, `hours`);
        expect(firstSlot.isSameOrAfter(minBookingTime)).toBe(true);
      }
    });

    it(`should handle short working hours`, () => {
      // Working only 2 hours
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 15:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 17:00`, TIMEZONE).valueOf(),
        pauseTimes: [],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Thirty),
      };

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 14:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        [],
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // Should have at least some slots
      expect(result.availableTimes.length).toBeGreaterThanOrEqual(0);
    });
  });

  describe(`✅ Regression Test: Snapshot current behavior`, () => {
    it(`should maintain consistent slot calculation (regression protection)`, () => {
      // This test protects against unintended changes
      const employee: EmployeeWorkingDayPure = {
        employeeId: 15,
        startWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 08:00`, TIMEZONE).valueOf(),
        endWorkingTimeMs: dayjs.tz(`${FUTURE_DATE} 20:00`, TIMEZONE).valueOf(),
        pauseTimes: [
          {
            startPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 12:00`, TIMEZONE).valueOf(),
            endPauseTimeMs: dayjs.tz(`${FUTURE_DATE} 13:00`, TIMEZONE).valueOf(),
          },
        ],
        advanceBookingTime: `00:00:00` as Time_HH_MM_SS_Type,
        timeslotInterval: Number(TimeslotIntervalEnum.Fifteen),
      };

      const blockers: NormalizedAppointmentPure[] = [
        {
          employeeId: 15,
          dateISO: FUTURE_DATE as Date_ISO_Type,
          startTimeMs: dayjs.tz(`${FUTURE_DATE} 09:00`, TIMEZONE).valueOf(),
          endTimeMs: dayjs.tz(`${FUTURE_DATE} 10:00`, TIMEZONE).valueOf(),
        },
      ];

      const currentTimeMs = dayjs.tz(`${FUTURE_DATE} 07:00`, TIMEZONE).valueOf();

      const result = calculateEmployeeDayAvailability(
        employee,
        blockers,
        currentTimeMs,
        FUTURE_DATE as Date_ISO_Type,
        FUTURE_DATE as Date_ISO_Type,
        `00:30:00` as Time_HH_MM_SS_Type,
      );

      // Snapshot test - if this fails, review changes carefully!
      expect(result.employeeId).toBe(15);
      expect(result.availableTimes.length).toMatchSnapshot(`regression-test-slot-count`);

      // Basic sanity checks
      expect(result.availableTimes.length).toBeGreaterThan(0);
      expect(result.availableTimes.length).toBeLessThan(100); // Reasonable upper bound
    });
  });
});
