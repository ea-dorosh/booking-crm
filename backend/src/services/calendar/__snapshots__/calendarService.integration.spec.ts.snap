// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Service Integration Tests âœ… Regression Test: Snapshot current behavior should maintain consistent slot calculation (regression protection): regression-test-slot-count 1`] = `3`;
